stm32: Add machine.CAN low-level driver

projectgus opened the pull request:
### Summary

This work follows on from #13149 and closes #12337. The goal is to have a `machine.CAN` API that can be used on multiple ports, and an implementation that works on stm32 port. The existing [pyb.CAN API](https://docs.micropython.org/en/latest/library/pyb.CAN.html) is being kept alongside this.

The `machine.CAN` API has changed substantially from the draft submitted in #13149. See [comment with more explanation about what happened](https://github.com/micropython/micropython/pull/13149#issuecomment-3653931146).

There's also one change in the old `pyb.CAN` API, which is that `recv()` now always returns a list rather than either a list or a tuple depending on how it is called. This allows using a new shared helper function that's added in one of the commits, but this change could also be pulled out at a small code size cost.

### Status

Documented functionality is implemented, and the supplied new tests all pass.

* [x] Decision about scoped Enums (see [comment](https://github.com/micropython/micropython/pull/18572#discussion_r2618189317)).
* [x] Implement `CAN.recv()` error flags and add test coverage.
* [x] Implement controller errors: `CAN.get_state()`, `CAN.State` enum, `CAN.IRQ_STATE` interrupt trigger.
* [x] Implement `CAN.get_counters()`
* [x] Implement `CAN.get_timings()`
* [x] Implement `CAN.reset()` and `CAN.restart()`, and/or simplify the API to remove one of these.
* [x] Implement `CAN.init(mode=x)` and add test coverage. (Note: Previously there was going to be a `CAN.mode()` function to change modes on the fly, but this isn't really supported by the controller so instead have to `deinit()` and then `init()` into a new mode. There is test coverage for doing this.
* [x] Verify `pyb.CAN` still works, using `multi_pyb_can` tests.
* [x] Look at possible code size reductions (note: didn't find a lot of low hanging fruit here, unfortunately.)

### Future PRs

* Use `target_wiring` in the `multi_extmod/machine_can_*` tests. Currently these test cases are all hard-wired for CAN instance 1.
* Create `can` & `aiocan` modules in micropython-lib. The goal of `machine.CAN` is to only implement the bare bones needed to make a higher level "Pythonic" CAN library on top.
* Implement CAN-FD support (the current PR works on STM32 CAN-FD controller peripherals, but only in Classic CAN mode).

### Testing

Ran `pyb.CAN` and new `machine.CAN` tests between pairs of `NUCLEO_H723ZG` board (fdcan peripheral), `NUCLEO_G474RE` board (simpler fdcan peripheral), and a `PYBDV11` (bxCAN peripheral).

Specific commands:

```
./run-tests.py -t a0 ports/stm32/pyb_can*.py extmod_hardware/machine_can*.py
```

and for each pairing of the three boards:

```
./run-multitests.py -p2 -t a0 -t a1 multi_extmod/machine_can_*.py multi_pyb_can/*.py
```

Also ran the all the same tests with assertions on for all three boards (setting `DEBUG=1` plus `COPT=-Os` for the G4 board). Some timeouts needed tweaking for these tests to pass.

In order to run `machine_can2.py` and `pyb_can2.py` tests on STM32H7, the same patch described in #18946 was applied.

### Trade-offs and Alternatives

* The biggest trade-off here was deciding to abandon the more ambitious API design from #13149 as it was too hard to implement properly. :laughing:. This pushes some more of the complexity into the eventual `micropython-lib` libraries, but it should be easier to implement all of it in Python now the Python-C interface is simpler.

projectgus committed:
extmod,docs: Add generic machine.CAN helpers & docs.

API is different to the original machine.CAN proposal, as numerous
shortcomings were found during initial implementation.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
stm32: Add machine.CAN implementation.

Implemented according to API docs in a parent comment.

Adds new multi_extmod/machine_can_* tests which pass when testing between
NUCLEO_G474RE, NUCLEO_H723ZG and PYBDV11.

This work was mostly funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

kwagyeman commented on docs/library/machine.CAN.rst:
```diff
@@ -0,0 +1,476 @@
+   Returns an irq object. If called with no arguments then a previously-configured
+   irq object is returned.
+
+.. method:: CAN.cancel_send(index)
```
Question, why support this at all? Generally, if you send a message on CAN you just expect it to be delivered...

projectgus committed:
py/objlist,stm32,esp32: Add helpers for creating/ensuring list args.

Simplifies the pattern of an optional arg which can be a list of at
least a certain length, otherwise one is lazily initialised.

Modify pyb.CAN and ESP-NOW APIs to use the helper. Note this changes
the return type of pyb.CAN.recv() from tuple to list.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

dpgeorge commented:
Thanks for putting this PR up!

I tested with 2x PYBV10 connected through some CAN tranceivers and the new tests added here all pass.

Testing the existing `multi_pyb_can` tests, the `multi_pyb_can/rx_callback.py` test fails with TIMEOUT.  The other test still passes.

I noticed there's not a simple multitest that just sends and recv's a few frames, without using IRQs.  I managed to get something simple working with a poll on `can.recv()` and it worked.  Maybe it's worth adding a simple test like that?  At least it would test that the CAN link is up and working, and serves as the simplest, working example.

dpgeorge commented on tests/multi_can/remote_req.py:
```diff
@@ -0,0 +1,74 @@
+from machine import CAN
```
Maybe these tests can go in `tests/multi_extmod/` and be prefixed with `machine_can_`?  Tests in that directory most likely require hardware connections anyway, and if you just want to run CAN tests then can do:
```
$ ./run-multitests.py -t a0 -t a1 multi_extmod/machine_can_*.py
```

dpgeorge commented on docs/library/machine.rst:
```diff
@@ -266,4 +266,5 @@ Classes
    machine.WDT.rst
    machine.SD.rst
    machine.SDCard.rst
+   machine.CAN.rst
```
Considering it's a communications bus, maybe put it after `machine.I2S.rst`?

dpgeorge commented on docs/library/machine.CAN.rst:
```diff
@@ -0,0 +1,476 @@
+   Construct a CAN controller object of the given id:
+
+   - ``id`` identifies a particular CAN controller object; it is board and port specific.
+   - ``bitrate`` is the initial bus bit rate in bits per second.
```
Why is `bitrate` explicitly listed here, and then also part of `init()`?  Would it be clearer to just leave it out of the argument list here?

dpgeorge commented on docs/library/machine.CAN.rst:
```diff
@@ -0,0 +1,476 @@
+   .. note:: Specific controller hardware may accept additional optional
+             parameters for hardware-specific features such as oversampling.
+
+.. method:: CAN.init_fd(bitrate, sample_point=None, sjw=None, tseg1=None, tseg2=None)
```
Does it make sense to combine this into `.init()` but with additional keyword arguments to set the timing?

I'm just thinking it's a little awkward to init FD mode with:
```py
can = machine.CAN(1, 500_000, CAN.Mode.SILENT)
can.init_fd(1_000_000)
can.mode(CAN.Mode.NORMAL)
```

Or maybe make it so the mode defaults to SILENT?

Not sure, just some ideas.

projectgus replied:
Have dropped this out of this PR for now. Because I also dropped support for changing mode without deinit first, enabling FD will probably end up being extra arguments on the `init()` function after all. But will submit in a separate PR.

dpgeorge commented on extmod/machine_can_port.h:
```diff
@@ -0,0 +1,139 @@
+// include is a no-op (as the file is included directly into
+// extmod/machine_can.c). However, including it anyway means that Language
+// Servers and IDEs can correctly analyse the machine_can.c file while the
+// developer is writing it.
```
Yes, this is a good approach.  Happy to (eventually) extend it to other extmod machine classes.

IhorNehrutsa commented on docs/library/machine.CAN.rst:
```diff
@@ -0,0 +1,563 @@
+
+These values are returned by :func:`CAN.get_state()` and reflect the error state of the CAN controller:
+
+.. data:: CAN.STATE_STOPPED
```
Please synchronize with pyb_can.c code
https://github.com/micropython/micropython/blob/eb7affcf0d02c1b83da61422880e7b88f504645f/ports/stm32/pyb_can.c#L815-L820

Must be 
```
data:: CAN.STOPPED
```
Same with others.

dpgeorge replied:
I think it's better to have the `STATE_` prefix.  That matches other (modern) APIs which group together common enums, eg `WLAN.IF_xxx`, `WLAN.SEC_xxx`, `WLAN.PM_xxx`, `Pin.IRQ_xxx` etc.

projectgus committed:
stm32: Add can_get_state() function, use from pyb.CAN.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
stm32: Implement index-aware STM32G4 FDCAN HAL TX functions.

These are oddly missing from the STM32G4 HAL, but the
reference manual describes being able to use them and
the implementations seem to work as expected.

Note that unlike STM32H7 it doesn't seem like we must use this approach,
because HAL_FDCAN_AddMessageToTxFifoQ() does seem to not have the issues
with priority inversion seen on the H7. However it's simpler to use the
same API for both...

Signed-off-by: Angus Gratton <angus@redyak.com.au>
Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
I think this is ready for review, I've updated the PR description with the current status. Still plenty of follow-up work (also listed in description).

The one thing I'm not 100% happy with is the code size hit of +4400 bytes on PYBV10. If I manually disable pyb.CAN and rebuild I get back 3864 bytes, so the new driver is not a lot bigger than the old driver (and has a more flexible API design) - but it's still a big hit.

projectgus commented:
> Do you want me to review?

@kwagyeman If you're able to then that would be great, thank you!

dpgeorge reviewed (commented):
There are some changes in the last commit which change the docs and generic code in `extmod/machine_can.c`.  I think those changes should be moved into the commit that adds the generic bindings + docs.

dpgeorge commented:
Regarding `CAN.set_filter()`: I see that passing in `[]` is a handy way to specify "no filters" ie "let everything through".  That feels a tiny bit counter intuitive, because it's not the degenerate case of passing in N filters when N=0 (the more filters you add to the list the more stuff gets let through, due to OR'ing, yet if I pass no filters I suddenly get everything).

To reduce code, size the code could simply not have the special case of "no filters" and then leave it up to the user to do `can.set_filter([0, 0, 0])` to let everything through.  After all, if it's intended to be a minimal API (and have a higher level `aiocan` which is more user friendly) then we can make such simplifications.

And I also feel that `can.set_filter(None)` is better suited to mean "no filters", ie accept everything.

So my proposal would be to make it so  `CAN.set_filters([])` adds no acceptance filters, ie doesn't let anything through (perhaps the method should be called `set_acceptance_filters()` but that's quite long).

Then could remove the `CAN.set_filters(None)` option altogether (to save code size, simplify the API).  Or take that to mean "accept everything".

projectgus commented on py/objlist.c:
```diff
@@ -520,3 +520,22 @@ mp_obj_t mp_obj_new_list_iterator(mp_obj_t list, size_t cur, mp_obj_iter_buf_t *
 }
+
+mp_obj_list_t *mp_obj_list_optional_arg(mp_obj_t arg_in, size_t min_len) {
+    if (arg_in == mp_const_none || MP_OBJ_TO_PTR(arg_in) == NULL) {
```
> Except maybe the use in machine_can_get_timings() for FDCAN mode, that may need tweaking.

Yeah, this is why I added this in the first place. Because `mp_obj_list_new()` gives you a list of NULL pointers back (memset to 0), if you call `mp_obj_list_optional_arg()` on a nested list then you have to special-case in NULL each time. So the diff looks like this:

```patch
diff --git i/extmod/machine_can.c w/extmod/machine_can.c
index 1d7de70059..c24ac3782a 100644
--- i/extmod/machine_can.c
+++ w/extmod/machine_can.c
@@ -608,6 +608,9 @@ static mp_obj_t machine_can_get_timings(size_t n_args, const mp_obj_t *args) {
     list->items[1] = MP_OBJ_NEW_SMALL_INT(self->sjw);
     list->items[2] = MP_OBJ_NEW_SMALL_INT(self->tseg1);
     list->items[3] = MP_OBJ_NEW_SMALL_INT(self->tseg2);
+    if (list->items[4] == MP_OBJ_NULL) {
+        list->items[4] = mp_const_none;
+    }
     #if MICROPY_HW_ENABLE_FDCAN
     mp_obj_list_t *fd_list = mp_obj_list_optional_arg(list->items[4], 4);
     fd_list->items[0] = mp_const_none;     // TODO: CAN-FD timings support
@@ -618,6 +621,9 @@ static mp_obj_t machine_can_get_timings(size_t n_args, const mp_obj_t *args) {
     #else
     list->items[4] = mp_const_none;
     #endif
+    if (list->items[5] == MP_OBJ_NULL) {
+        list->items[5] = mp_const_none;
+    }
     list->items[5] = machine_can_port_get_additional_timings(self, n_args > 1 ? list->items[5] : mp_const_none);
 
     return MP_OBJ_FROM_PTR(list);
diff --git i/py/objlist.c w/py/objlist.c
index 479a52c383..bb8f54eb6b 100644
--- i/py/objlist.c
+++ w/py/objlist.c
@@ -522,7 +522,7 @@ mp_obj_t mp_obj_new_list_iterator(mp_obj_t list, size_t cur, mp_obj_iter_buf_t *
 }
 
 mp_obj_list_t *mp_obj_list_optional_arg(mp_obj_t arg_in, size_t min_len) {
-    if (arg_in == mp_const_none || MP_OBJ_TO_PTR(arg_in) == NULL) {
+    if (arg_in == mp_const_none) {
         return MP_OBJ_TO_PTR(mp_obj_new_list(min_len, NULL));
     } else {
         return mp_obj_list_ensure(arg_in, min_len);
```

(This code path gets hit if you call `get_timings(None)` the first time around the loop, to allocate the data structures.)

I am happy to apply this version, but I wanted to point this out as it's a little bit of a trap if missed.

> There's one use of mp_obj_list_optional_arg() that passes NULL as this arg, which can easily be changed to pass mp_const_none. Other uses of it already do it that way (pass in mp_const_none).

Do you mean the explicit case in the coverage test, or is there another real usage of this that I missed?

projectgus replied:
Although can maybe make a smaller check like:

```c
if (n_args > 1 and args[1] == mp_const_none) {
  n_args = 0;
}
```

which is a bit obscure, but it'd cover this case.

projectgus commented on ports/stm32/fdcan.c:
```diff
@@ -362,8 +505,67 @@ int can_receive(FDCAN_HandleTypeDef *can, can_rx_fifo_t fifo, FDCAN_RxHeaderType
+    uint32_t psr;
+
+    if (can->State != HAL_FDCAN_STATE_BUSY) {
+        return CAN_STATE_STOPPED;
```
(Commented added.)

projectgus commented:
Made a couple of last-minute tweaks where docs and behaviour had diverged, and added a couple more examples to the docs.

Have now re-tested on G4, H7 and F4 (PYBV11) with all permutations, same as the initial PR.

dpgeorge commented on extmod/machine_can.c:
```diff
@@ -0,0 +1,757 @@
+            size_t arg_len;
+            mp_obj_get_array(filter, &arg_len, &args);
+            if (arg_len != 3) {
+                mp_raise_ValueError(MP_ERROR_TEXT("Filter items must have length 3"));
```
Please make first word lowercase.

dpgeorge reviewed (dismissed):
This looks really good now.  I tested again on PYBV10 <-> PYBV10 and all the tests pass.

Just a few minor comments, then this will be ready to merge.

projectgus commented on docs/library/machine.CAN.rst:
```diff
@@ -0,0 +1,634 @@
+     can.set_filters(((0x300, 0x700, 0),
+                      (0x50700, 0x1FFF_FFFF, CAN.FLAG_EXT_ID)))
+
+.. data:: CAN.filters_max
```
Oops, good catch! Have renamed to `CAN.FILTERS_MAX` (as it's a constant) and added test coverage that it's exposed.

dpgeorge commented on extmod/machine_can.c:
```diff
@@ -0,0 +1,751 @@
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2024 Angus Gratton
```
Probably worth updating the copyright year to 2024-2026.

chrismas9 commented:
I have raised an issue for pyb.CAN. I only raised it here because machine.CAN maybe should be tested for the same issue.

https://github.com/micropython/micropython/issues/18922  STM32 pyb.CAN: Intantiating second instance corrupts settings of first instance.

dpgeorge reviewed (approved):
Thanks for updating and addressing all the feedback.  Much appreciated.

I tested on:
- PYBV10 standalone (tests bxCAN, CAN1 and CAN2)
- PYBD_SF2 standalone (tests bxCAN, CAN1 only)
- NUCLEO_H723ZG (tests FDCAD, CAN1 only)
- PYBV10 - PYBV10 multi tests

Everything I tested passes.

iabdalkader commented on ports/stm32/mpconfigport.h:
```diff
@@ -122,6 +122,14 @@
 #define MICROPY_PY_MACHINE_BITSTREAM (1)
 #endif
+#ifndef MICROPY_PY_MACHINE_CAN
+#ifdef MICROPY_HW_CAN1_TX
```
@projectgus @dpgeorge Is there any reason why this is not enabled for `CAN2_TX` too? I have boards that use CAN2, and I just added a `|| defined(CAN2_TX)`, am I missing something?

dpgeorge replied:
I think originally the thought was that if the board has two CAN peripherals then both CAN1 and CAN2 will be enabled.

But I guess it's possible that only CAN2 is enabled.  So changing as you suggest makes sense.
